tests/how_to_test_full_app.rs

```diff
@@ -0,0 +1,113 @@
+use bevy::prelude::*;
+
+const DEFAULT_MANA: u32 = 10;
+
+#[derive(Component)]
+struct Player {
+    mana: u32,
+}
+
+fn window_title_system(mut windows: Query<&mut Window>) {
+    for (index, mut window) in windows.iter_mut().enumerate() {
+        window.title = format!("This is window {index}!");
+    }
+}
+
+fn spawn_player(mut commands: Commands) {
+    commands.spawn(Player { mana: DEFAULT_MANA });
+}
+
+fn spell_casting(mut player: Query<&mut Player>, keyboard_input: Res<Input<KeyCode>>) {
+    if keyboard_input.just_pressed(KeyCode::Space) {
+        let Ok(mut player) = player.get_single_mut() else {
+            return;
+        };
+
+        if player.mana > 0 {
+            player.mana -= 1;
+        }
+    }
+}
+
+fn create_test_app() -> App {
+    let mut app = App::new();
+
+    // Note how we use `MinimalPlugins` instead of `DefaultPlugins`.
+    // This is what allows the test to run without a window, or real user input.
+    app.add_plugins(MinimalPlugins);
+    // Inserting a KeyCode input resource allows us to inject keyboard inputs,
+    // as if the user had pressed them.
+    app.insert_resource(Input::<KeyCode>::default());
+
+    // Spawning a fake window allows testing systems that require a window.
+    app.world.spawn(Window::default());
+
+    app
+}
+
+fn add_game_systems(app: &mut App) {
```

Comment:
This should use a plugin: it's extremely useful to refactor your game such that some plugins can be added directly to the test setup.

Reply:
Good point, I have changed it into a plugin.